02-peer-protocol.md

```diff
@@ -776,6 +788,83 @@ satoshis, which is possible if `dust_limit_satoshis` is below 546 satoshis).
 No funds are at risk when that happens, but the channel must be force-closed as
 the closing transaction will likely never reach miners.
 
+### Closing Negotiation: `closing_complete` and `closing_sig`
+
+Once shutdown is complete, the channel is empty of HTLCs, there are no commitments
+for which a revocation is owed, and all updates are included on both commitments,
+the final current commitment transactions will have no HTLCs.
+
+Each peer says what fee it will pay, and the other side simply signs that transaction.  The only complexity comes from allowing each side to omit its own output should it be uneconomic.
+
+This process will be repeated every time a `shutdown` message is received, which allows re-negotiation.
+
+1. type: 40 (`closing_complete`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u64`:`fee_satoshis`]
+   * [`u8`: `has_closer_output`]
+   * [`signature`:`signature_with_closee_output`]
+   * [`signature`:`signature_without_closee_output`]
+
+1. type: 41 (`closing_sig`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u8`: `closee_output`]
+   * [`signature`:`signature`]
+
+#### Requirements
+
+Note: the details and requirements for the transaction being signed are in [BOLT 3](03-transactions.md#closing-transaction)).
+
+Both nodes:
+  - After a `shutdown` has been received, AND no HTLCs remain in either commitment transaction:
+    - SHOULD send a `closing_complete` message.
+
+The sender of `closing_complete` (aka. "the closer"):
+  - MUST set `fee_satoshis` to a fee less than or equal to its outstanding balance, rounded down to whole satoshis.
+  - SHOULD set `has_closer_output` to 0 if it considers its own remaining balance to be uneconomic.
+  - Otherwise MUST set `has_closer_output` to 1.
+  - If it sets `has_closer_output` to 1:
+    - MUST set `signature_with_closee_output` to a valid signature of a transaction with both closer and closee outputs.
+	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only a closer output.
+  - Otherwise: (`has_closer_output` is 0):
+    - MUST set `signature_with_closee_output` to a valid signature of a transaction with only the closee output.
+	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.
```

Comment:
Might be missing something but I don't think standardness will allow this case into mempool/relay.

1 native segwit input with 1 OP_RETURN output is ~61 bytes of non-witness data.

Bitcoin Core would allow 65+ starting with Bitcoin Core 25.0: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-25.0.md#notable-changes

but sub-64 is still not allowed.

Reply:
It's defined in BOLT 3 which covers the actual tx construction...

Reply:
If you're in a cooperative close scenario, but you've decided you're happy to abandon all your channel funds, why not donate them to your counterparty rather than miners? ie, add what would have been your balance to your counterparty's output value. If your counterparty then wants to move some of that to fees, they can re-negotiate the cooperative close and do an rbf?

Adding null output handling doesn't seem in line with the goal of making things simpler and less special case-y.

If you are signing a "send all my funds to fees" tx, you should probably be doing it in an aggregatable way (`ANYONECANPAY|NONE` or `ANYONECANPAY|ALL`) so that in the event of a bunch of channels being closed because block space is scarce and expensive, the 10+ byte outputs of all those channels can be merged.

Reply:
But why add any of this complexity in the first place? If we're in that scenario, that means the balance of the closer was very low (otherwise they wouldn't omit their output) which means that the balance on the closee side is large enough that they'll never want to omit it (it doesn't make sense to allow channels with a total capacity that is ~dust).

We should only send a signature for a transaction that contain the closee output, and omit the case where the transaction doesn't contain any real output, don't we?

Reply:
> why not donate them to your counterparty rather than miners?

Adding financial incentives to goad coop closes seems suboptimal.

> We should only send a signature for a transaction that contain the closee output, and omit the case where the transaction doesn't contain any real output, don't we?

The non-initiator can burn their dust later if they so choose, and in any manner they choose, without having to specify exactly how. I think that's a vote in favor of it.

Reply:
> Adding additional financial incentives to goad coop closes seems suboptimal.

I think the logic here should be the same as for dust: "if this output will never be worth spending, don't create it". If done that way, there's no opportunity for goading -- first, you should never allow your balance to drop to that level, and in the event that it somehow does (your channel was down while fees were rising, you changed your configuration and increased your dust value), you simply unconditionally abandon the channel when it's closed.

If your threshold for a spendable output is X, you estimate a cooperative close fee as F, you estimate a unilateral close fee as U (U>F), and your balance is B, then the logic is:

 * if `B > X`, reject any forward that would drop B below X. (or, for balance privacy, if `B < X + htlc_maximum_msat`, reject any forward to this channel)
 * if `B < X`, always give the balance to your counterparty.
 * if `B < X+U` consider rebalancing to increase B above X+U (or, if supported, rebalancing(draining) the channel to 0 and closing it).
 * if `0 < B < X+F`, don't automatically close the channel. if the user insists anyway, give the balance to your counterparty.
 * if `X+F < B < X+U` and rebalancing isn't possible, consider proposing a mutual close.

If you initiate a mutual close in the last case, you get an output with (B-F > X) funds, and can choose a low fee rate that will be slow to confirm (it's only a small output anyway, so you probably don't care very much when it's confirmed), leaving your counterparty to feebump if they want access to their funds sooner.

(Also, burning funds to fees doesn't remove the incentive to game a channel close, it at best moves the incentive from your counterparty to miners, and, if your counterparty was going to pay lots of fees anyway, just funds your counterparty less directly)

Reply:
Two points:

- One day, splice-to-close will be a thing, and you'll get to carefully craft your close tx to do arbitrary things.
- I don't expect to ever see a "no output" close on mainnet, but we have to cover it.

Reply:
> I don't expect to ever see a "no output" close on mainnet, but we have to cover it.

mainnet is the only place where fees might plausibly render an entire channel to be dust making burning the entire balance to fees sensible. If the feature's not useful there, it isn't useful anywhere.

In the event that a pair of nodes really do want to abandon a channel balance, they can manually specify a scriptPubKey of  "OP_RETURN [padding]" or "p2pkh[G]" to either permanently burn the funds (ultrasound money!) or donate to some future miner that's got some free space in a block. Anyway, if you want to make your simple close unnecessarily complicated, I guess that's your call, so I won't harp on this further...